In [70]:
import pandas as pd
import numpy as np

np.random.seed(1212)

import keras
from keras.models import Model
from keras.layers import *
from keras import optimizers

In [71]:
df_train = pd.read_csv('train.csv')
# df_test = pd.read_csv('../input/test.csv')

In [72]:
data = np.array(df_train)

data=data.T

y_train = data[0]

X_train = data[1:].T

In [73]:
print(X_train.shape)

(42000, 784)


In [74]:
print(y_train.shape)

(42000,)


## Splitting into training and validation dataset


## Data cleaning, normalization and selection


In [75]:
# print((min(X_train[1]), max(X_train[1])))

As the pixel intensities are currently between the range of 0 and 255, we proceed to normalize the features, using broadcasting. In addition, we proceed to convert our labels from a class vector to binary One Hot Encoded


In [76]:
# Feature Normalization 
X_train = X_train.astype('float32')
X_train /= 255

# Convert labels to One Hot Encoded
num_digits = 10
y_train = keras.utils.to_categorical(y_train, num_digits)

In [77]:
# Printing 2 examples of labels after conversion
# print(y_train[0]) # 2
# print(y_train[3]) # 7

## Model Fitting


In [78]:
# Insert Hyperparameters
learning_rate = 0.1
training_epochs = 20
batch_size = 100
sgd = optimizers.SGD(lr=learning_rate)

c:\Users\aryap\anaconda3\envs\Tensorflow\lib\site-packages\keras\optimizers\optimizer_v2\gradient_descent.py:111: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.
  super().__init__(name, **kwargs)


In [79]:
# Input Parameters
n_input = 784 # number of features
n_hidden_1 = 300
n_hidden_2 = 100
n_hidden_3 = 100
n_hidden_4 = 200
num_digits = 10

In [80]:
Inp = Input(shape=(784,))
x = Dense(n_hidden_1, activation='relu', name = "Hidden_Layer_1")(Inp)
x = Dropout(0.3)(x)
x = Dense(n_hidden_2, activation='relu', name = "Hidden_Layer_2")(x)
x = Dropout(0.3)(x)
x = Dense(n_hidden_3, activation='relu', name = "Hidden_Layer_3")(x)
x = Dropout(0.3)(x)
x = Dense(n_hidden_4, activation='relu', name = "Hidden_Layer_4")(x)
output = Dense(num_digits, activation='softmax', name = "Output_Layer")(x)

In [81]:
# Our model would have '6' layers - input layer, 4 hidden layer and 1 output layer
model4 = Model(Inp, output)
model4.summary() # We have 297,910 parameters to estimate

Model: "model_3"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 input_4 (InputLayer)        [(None, 784)]             0         
                                                                 
 Hidden_Layer_1 (Dense)      (None, 300)               235500    
                                                                 
 dropout_9 (Dropout)         (None, 300)               0         
                                                                 
 Hidden_Layer_2 (Dense)      (None, 100)               30100     
                                                                 
 dropout_10 (Dropout)        (None, 100)               0         
                                                                 
 Hidden_Layer_3 (Dense)      (None, 100)               10100     
                                                                 
 dropout_11 (Dropout)        (None, 100)               0   

In [82]:
model4.compile(loss='categorical_crossentropy',
              optimizer='adam',
              metrics=['accuracy'])

In [83]:
history = model4.fit(X_train, y_train,
                    batch_size = batch_size,
                    epochs = training_epochs)

Epoch 1/20
420/420 [==============================] - 3s 4ms/step - loss: 0.5183 - accuracy: 0.8357
Epoch 2/20
420/420 [==============================] - 2s 5ms/step - loss: 0.2041 - accuracy: 0.9411
Epoch 3/20
420/420 [==============================] - 2s 4ms/step - loss: 0.1543 - accuracy: 0.9545
Epoch 4/20
420/420 [==============================] - 3s 6ms/step - loss: 0.1307 - accuracy: 0.9619
Epoch 5/20
420/420 [==============================] - 3s 7ms/step - loss: 0.1076 - accuracy: 0.9694
Epoch 6/20
420/420 [==============================] - 2s 6ms/step - loss: 0.1024 - accuracy: 0.9709
Epoch 7/20
420/420 [==============================] - 2s 5ms/step - loss: 0.0922 - accuracy: 0.9730
Epoch 8/20
420/420 [==============================] - 2s 6ms/step - loss: 0.0832 - accuracy: 0.9759
Epoch 9/20
420/420 [==============================] - 2s 6ms/step - loss: 0.0750 - accuracy: 0.9789
Epoch 10/20
420/420 [==============================] - 2s 5ms/step - loss: 0.0713 - accuracy: 0.9791

In [84]:
model4.save('Digit_Recognition_model.keras')

In [86]:
loaded_model = tf.keras.models.load_model('Digit_Recognition_model.keras')

In [88]:
loaded_model.summary()

Model: "model_3"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 input_4 (InputLayer)        [(None, 784)]             0         
                                                                 
 Hidden_Layer_1 (Dense)      (None, 300)               235500    
                                                                 
 dropout_9 (Dropout)         (None, 300)               0         
                                                                 
 Hidden_Layer_2 (Dense)      (None, 100)               30100     
                                                                 
 dropout_10 (Dropout)        (None, 100)               0         
                                                                 
 Hidden_Layer_3 (Dense)      (None, 100)               10100     
                                                                 
 dropout_11 (Dropout)        (None, 100)               0   

With a validation score of close to 98%, we proceed to use this model to predict for the test set.
